lib/Dialect/Triton/IR/Ops.cpp

```diff
@@ -688,29 +688,44 @@ LogicalResult ExpandDimsOp::canonicalize(ExpandDimsOp op,
                                                  splat.getOperand());
     return mlir::success();
   }
-  // expand_dims(broadcast) -> broadcast(expand_dims)
+  // expand_dims(broadcast(x)) -> broadcast(expand_dims(x))
   //
-  // On it's own this doesn't do much, but consider
+  // On its own this doesn't do much, but consider
   //    broadcast(expand_dims(broadcast))
   // -> broadcast(broadcast(expand_dims))
   // -> broadcast(expand_dims)
+  //
+  // Note we can't do this if x is a scalar, because expandDims requires a
+  // tensor input.  (We could change it to broadcast(expand_dims(splat(x))), but
+  // it's unclear that's better.)  broadcast currently accepts scalar inputs,
+  // but it probably shouldn't (just use splat instead).
```

Comment:
I'm confused with this comment. What I'm understanding is that we don't know whether `broadcast(expand_dims(splat(x)))` is good, so we'll resort to allowing `broadcast` to take scalar, but you don't think this is the right way to go and instead should use `splat`, which we don't know if it's better? I'm just trying to understand the decision, let me know if there was a thread somewhere so I can catch up on.

Reply:
Oh I see, you're saying we should use `splat(x)` instead of `broadcast(expand_dims(splat(x)))` ? That makes sense

Reply:
Yeah, ultimately it should become splat, although exactly what series of one or more rewrite rules it uses to get there is unclear.